File: 3sum/main.go
package threesum

import (
	"sort"
)

func ThreeSum(nums []int) [][]int {
	sort.Ints(nums)
	var result [][]int
	var prevA int

	for i, num := range nums {
		if num == prevA && i != 0 {
			continue
		}
		l, r := i+1, len(nums)-1
		// init prevB like this because this initial will never includes in nums (nums is sorted so I can do this)
		prevB := nums[0] - 1

		for l < r {
			if nums[l] == prevB {
				l++
				continue
			}
			tmp := nums[l] + nums[r]
			if tmp == -num {
				result = append(result, []int{num, nums[l], nums[r]})
				prevB = nums[l]
				l++
				r--
			} else if tmp < -num {
				l++
			} else {
				r--
			}
		}
		prevA = num
	}
	return result
}

File: 3sum/c++/main.cpp
#include <algorithm>
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<int>> threeSum(vector<int> &nums) {
    sort(nums.begin(), nums.end(), [](int a, int b) { return a < b; });
    vector<vector<int>> result;
    int prevA;

    for (int i = 0; i < nums.size(); i++) {
      int num = nums[i];
      if (num == prevA && i != 0) {
        continue;
      }
      int l = i + 1;
      int r = nums.size() - 1;
      int prevB = nums[0] - 1;

      while (l < r) {
        if (nums[l] == prevB) {
          l++;
          continue;
        }
        int tmp = nums[l] + nums[r];
        if (tmp == -num) {
          result.push_back({
              num,
              nums[l],
              nums[r],
          });
          prevB = nums[l];
          l++;
          r--;
        } else if (tmp < -num) {
          l++;
        } else {
          r--;
        }
      }
      prevA = num;
    }
    return result;
  }
};
